### DISCUSSION TITLE
Spurious axons, half axons, partial myelin axons

### DISCUSSION THREAD

**Discussion by akraskovSMDB** [neuropoly/axondeepseg#656] [Show and tell]
Hi,

we got a problem with falsely detected axons within a histological artifact of an image.
Original LM image is shown below 

![image](https://user-images.githubusercontent.com/25688017/175294308-8a9239d4-8ec4-4db7-8543-4020a5aea8a5.png)

Superposition of the original image with automatically segmented image showing axon centres and diameters.
![image](https://user-images.githubusercontent.com/25688017/175293760-13e08689-d11b-468a-868b-a0e4035ee3a9.png)

Problematic axons are within a big void. Part of this void was falsely segmented as inner axon mask. This probably led to detect axons 968, 1119, 1128, 1170, 1285, 1419, 1489 on the edges of this void. But all of these axons have "myelin" only on one side (void edges side) and nothing on the other side (inside void). I accept that this myelin is falsely segmented but difficult for the model to ignore due to the similar intensity. I was hoping that because myelin is segmented only one side of these "axons" they will be ignored during morphometrics analysis.

 Is there a way to ignore axon with only partially segmented myelin (in this case more around half) and exclude falsely detected axons? Can some measure of myelin integrity/continuity be defined?

Thank you

**Comment by hermancollin** [comment]
Hi @akraskovSMDB 
Would it be possible to share the original axonmyelin mask? I'm having trouble seeing the inner axon mask you are talking about.

> Is there a way to ignore axon with only partially segmented myelin (in this case more around half) and exclude falsely detected axons? Can some measure of myelin integrity/continuity be defined?

That would indeed be a useful feature. If we could define such a metric, it could be added as a boolean value in the morphometrics file to indicate whether or not the myelin is "complete".

In the meantime, we are working on a feature to easily remove axons from a mask. Do you think this could also help you?

**Reply by akraskovSMDB** [reply]
Here is a fragment of the mask roughly corresponding to the region with bad axons.
![image](https://user-images.githubusercontent.com/25688017/175326278-38eae04b-f232-4574-8c8c-49e2dafd1929.png)

> In the meantime, we are working on a feature to easily remove axons from a mask. Do you think this could also help you?
I don't know, I need to understand how it will work? What is the criteria to remove an axon from a mask?

**Reply by akraskovSMDB** [reply]
Here is a mask for axons and myelin
![image](https://user-images.githubusercontent.com/25688017/175327150-54a14182-b8f7-4b06-aa8e-e024f8baa3ca.png)

**Reply by hermancollin** [reply]
Ah yes, I see what you mean.

I will say that this is an issue we often encounter, especially with images like the one you provided where an histological artifact vaguely ressembling an axon or myelin is falsely segmented. This will often result in multiple "axon" object with myelin only on one side. So this is a use-case we would definitely want to cover more appropriately. At a first glance, I think a simple morphological operation on the axonmyelin object could easily detect this kind of artifact. I will look into this.
I'm thinking of a morphometric flag we would set to `true` if an axon isn't entirely contained inside its corresponding myelin.

> > In the meantime, we are working on a feature to easily remove axons from a mask. Do you think this could also help you?
> 
> I don't know, I need to understand how it will work? What is the criteria to remove an axon from a mask?

I can briefly explain the criteria if you are familiar with image analysis. To perform instance segmentation, so that every axon is individually "separated" from the others for morphometrics analysis, we use a watershed algorithm to transform the semantic segmentation mask that the model outputs into an instance segmentation. The tool our team is working on will remove an axon based on this postprocessed instance segmentation. So, intuitively, the result can be hard to describe because the watershed algorithm will sometimes do unpredictable stuff. With that being said, most of the problematic objects on your image could be easily removed with this tool.

**Reply by Stoyan-I-A** [reply]
In response to @hermancollin, I believe @akraskovSMDB wants to know what will be used to determine if a specific axon would be selected to be removed automatically. 
And we don't have a definitive answer to this. We think we could automatically filter out axons that are too small (e.i. their area is a very small percentage of the total image). Or we could allow to have multiple criteria (exemple: a user could choose au automatically remove axons that are touching the border of the image).

**Reply by akraskovSMDB** [reply]
> I will say that this is an issue we often encounter, ...  At a first glance, I think a simple morphological operation on the axonmyelin object could easily detect this kind of artifact. I will look into this. I'm thinking of a morphometric flag we would set to `true` if an axon isn't entirely contained inside its corresponding myelin.

This would be great. Do you know already what specific "simple morphological operation" you want to you use or still need to develop it?

> I can briefly explain the criteria if you are familiar with image analysis. To perform instance segmentation, so that every axon is individually "separated" from the others for morphometrics analysis, we use a watershed algorithm to transform the semantic segmentation mask that the model outputs into an instance segmentation. The tool our team is working on will remove an axon based on this postprocessed instance segmentation. So, intuitively, the result can be hard to describe because the watershed algorithm will sometimes do unpredictable stuff. With that being said, most of the problematic objects on your image could be easily removed with this tool.

It seems to be a more advanced technique which aims to solve a more difficult problem. As you say it probably will help to solve my problem as well but I presume it will need more time to implement, therefore the first solution seems to be more appropriate and quicker

Thank you

**Reply by akraskovSMDB** [reply]
> In response to @hermancollin, I believe @akraskovSMDB wants to know what will be used to determine if a specific axon would be selected to be removed automatically. And we don't have a definitive answer to this. We think we could automatically filter out axons that are too small (e.i. their area is a very small percentage of the total image). Or we could allow to have multiple criteria (exemple: a user could choose au automatically remove axons that are touching the border of the image).

I understand that the watershed algorithm will not allow to set specific criteria on individual axons which is fine. Removal of small axons can be done post-hoc just based on morphometrics output, similarly rejecting axons touching boundaries already implemented in response to my previous request(s) (bounding box output is still not approved hopefully will be soon)

**Reply by hermancollin** [reply]
> This would be great. Do you know already what specific "simple morphological operation" you want to you use or still need to develop it?

@akraskovSMDB hm well it wouldn't even be a morphological operation. All we need is to detect if the axon is adjacent to the background class. Normally, the axon class should be entirely contained inside its myelin without any "background" inbetween. We will keep you posted with these developments.

> (bounding box output is still not approved hopefully will be soon)

Apologies, it should be merged by now (we missed our bi-weekly meeting because of a national holiday). I'll notify the team to review the code ASAP.

edit: @akraskovSMDB ah, I see @mathieuboudreau already took care of it. Feel free to tell us if you have any issue with this new feature (you're definitely the first person who will use it). Cheers.

**Reply by akraskovSMDB** [reply]
> > This would be great. Do you know already what specific "simple morphological operation" you want to you use or still need to develop it?
> 
> @akraskovSMDB hm well it wouldn't even be a morphological operation. All we need is to detect if the axon is adjacent to the background class. Normally, the axon class should be entirely contained inside its myelin without any "background" inbetween. We will keep you posted with these developments.

I noticed that sometimes inner axon space within myelin is not completely labelled as axon, i.e. it might have some background inclusions, may be one can allow small percentage of background within myelin. Or maybe even better way would be to check for completeness of the myelin around any axon label (not sure how it can be implemented).


> 
> > (bounding box output is still not approved hopefully will be soon)
Thank you, will test it and report in case of any problems

**Comment by hermancollin** [comment]
@akraskovSMDB you are right, this is a segmentation artifact we often encounter (and could be easily postprocessed given the certainty that the myelin is actually complete). For this reason, checking if the axon is entirely contained in its associated myelin isn't exactly what we are looking for.

> Or maybe even better way would be to check for completeness of the myelin around any axon label (not sure how it can be implemented).

What I'm thinking is the following algorithm: 
```
1. take N values between 0 and 360 degrees (N could be adjusted to minimize computation-time)
2. for every angle, draw a line that intersects the axon center with a slope computed from the aforementioned angle
3. check if the line intersects the myelin class for the given axon
```

And I reckon doing this for every axon would give an excellent approximation for "completeness" of the myelin, given that the angle granularity is appropriate for the resolution of the image.

**Reply by akraskovSMDB** [reply]
> @akraskovSMDB you are right, this is a segmentation artifact we often encounter (and could be easily postprocessed given the certainty that the myelin is actually complete). For this reason, checking if the axon is entirely contained in its associated myelin isn't exactly what we are looking for.
> 
> > Or maybe even better way would be to check for completeness of the myelin around any axon label (not sure how it can be implemented).
> 
> What I'm thinking is the following algorithm:
> 
> ```
> 1. take N values between 0 and 360 degrees (N could be adjusted to minimize computation-time)
> 2. for every angle, draw a line that intersects the axon center with a slope computed from the aforementioned angle
> 3. check if the line intersects the myelin class for the given axon
> ```
> 
> And I reckon doing this for every axon would give an excellent approximation for "completeness" of the myelin.

Sounds good to me, one can even export (if an option is specified) proportion of of radii intersecting myelin into morphometrics output. 

As an additional feature which can be implemented during the same completeness check, it would be interesting to see "evenness" of the myelin thickness. I've got some examples, see below, where myelin thickness is not uniform around the axon (I don't think it is anatomical, probably more related to the tissue processing). At many points myelin sheaths look to be separated from each other increasing average myelin thickness measurement. For a given example, the "true" myelin thickness should be defined trough thinnest part (~4 o'clock direction). Suggested by you procedure (samping N directions) should allow to do this?


![image](https://user-images.githubusercontent.com/25688017/176254267-0965a363-735a-46ec-9447-b9442c145eb4.png)

**Reply by hermancollin** [reply]
> Sounds good to me, one can even export (if an option is specified) proportion of of radii intersecting myelin into morphometrics output.

Would that be useful for you?

> At many points myelin sheaths look to be separated from each other increasing average myelin thickness measurement. For a given example, the "true" myelin thickness should be defined trough thinnest part (~4 o'clock direction). Suggested by you procedure (samping N directions) should allow to do this?

It would indeed be possible to check for this as well, although a little bit more complex (because we not only need to find a point intersecting the axon center and the myelin, but also the closest and the farthest points from the axon center). What would you suggest for this "evenness" metric? Simply mean + standard deviation?

**Reply by akraskovSMDB** [reply]
> > Sounds good to me, one can even export (if an option is specified) proportion of of radii intersecting myelin into morphometrics output.
> 
> Would that be useful for you?

Yep, I will use it as a quality measure, and depending on how strict I want to be with my analysis I will set a threshold, .e.g if it is above 95%, it is likely to be "a good" axon and myelin.

> 
> > At many points myelin sheaths look to be separated from each other increasing average myelin thickness measurement. For a given example, the "true" myelin thickness should be defined trough thinnest part (~4 o'clock direction). Suggested by you procedure (samping N directions) should allow to do this?
> 
> It would indeed be possible to check for this as well, although a little bit more complex (because we not only need to find a point intersecting the axon center and the myelin, but also the closest and the farthest points from the axon center).
Correct
>
> What would you suggest for this "evenness" metric? Simply mean + standard deviation?

Yep, these will be useful, but I would add min value as well, since I believe that this value is a true myelin thickness; to avoid potential problems with non-complete myelin a 10% percentile might be better, or to be even more general introduce a parameter specifying a desired percentile, which will give an option for min, 10%,  median etc

**Comment by hermancollin** [comment]
I want to make sure we are on the same page before implementing all of this.

## feature 1: myelin completeness check
Leveraging the algorithm described above, check for myelin "completeness" (or "integrity") and output boolean value in morphometrics. Partially segmented myelin (i.e. like in the first mask posted here) should be identified as incomplete.

## feature 2: myelin evenness analysis
During the step where we check for myelin completeness, we will collect other measures that will give an idea of how even the myelin thickness is. Measures that we should save are:
- myelin thickness mean
- myelin thickness standard deviation
- myelin thickness 10th percentile

## feature 3: proportion of radii intersecting the myelin
So, this one I'm not really sure about. Essentially, what you were asking here:
> one can even export (if an option is specified) proportion of of radii intersecting myelin into morphometrics output.

was to compute the g-ratio accurately rather than using approximations like we do in ADS. This is interesting to me. I would be curious to see the difference between the approximation we use now and the "harder" approximation obtained from collecting ratios around the axon and averaging them (we could also add more measures related to this like standard deviation, etc.)

All in all, it is obvious to me that this will require to add a new morphometrics flag, because the computation will undoubtedly be longer and I would like to keep this as an option for the user. 
@akraskovSMDB comments? Something missing?

**Reply by akraskovSMDB** [reply]
Thank you for summarising.
In my view Feature 3 is a part of Feature 1, proportion of radii intersecting myelin is just a quantification of completeness/integrity, if proportion is 1 than myelin is complete. Proportions close to 1 (but not 1) might be still considered as ok, if the segmentation is not perfect.

I think your comment about the computation of g-ratio is more relevant for Feature 2.  Min vs 10th percentile vs mean will give different g-ratios (decreasing). Mean value should give closest g-ratio to your current method of estimation. A caveat to this is that any variation in myelin thickness is probably due to the tissue preprocessing rather than genuine axon properties. I don't know any anatomical/physiological reason for variable myelin thickness. Do you? BUT I still think that this feature is useful since experimental data is never perfect and it is good to have a way to compensate for some of the tissue preparation problems.  

My summary of the features will as following
Feature 1
input: 
--m-integrity [N] , where N is an optional parameter specifying number of radii to check, default 100
output:
in morphometrics files add two columns: one is a proportion of radii intersecting myelin, and second is boolean which is 1 if proportion is 1 (a bit redundant)

Feature 2:
input:
--m-evenness [P], where P is optional and is a percentile of myelin thickness distribution around axon (default 10)
output:
in morphometrics files add _three_ columns: mean and std of myelin thickness distribution, plus Pth percentile of this distribution (if P is set to 0, i.e. a min value is requested, zero value will mean that axon is not complete, sanity check for features 1 and 2)

I guess both options can be combined into one flag with two optional parameters, -m [N] [P]

It is obviously up to you how you set it up, thank you for consideration.

**Reply by hermancollin** [reply]
> I think your comment about the computation of g-ratio is more relevant for Feature 2. Min vs 10th percentile vs mean will give different g-ratios (decreasing). Mean value should give closest g-ratio to your current method of estimation. A caveat to this is that any variation in myelin thickness is probably due to the tissue preprocessing rather than genuine axon properties. I don't know any anatomical/physiological reason for variable myelin thickness. Do you? BUT I still think that this feature is useful since experimental data is never perfect and it is good to have a way to compensate for some of the tissue preparation problems.
> 
> My summary of the features will as following Feature 1 input: --m-integrity [N] , where N is an optional parameter specifying number of radii to check, default 100 output: in morphometrics files add two columns: one is a proportion of radii intersecting myelin, and second is boolean which is 1 if proportion is 1 (a bit redundant)
> 

Ah, I misunderstood what you meant by "proportion of radii intersecting myelin". Yes, this makes a lot of sense. Furthermore, the boolean value would be totally useless if this value was provided. I think this feature should output the percentage value instead of the boolean. So this part would only add one column in the morphometrics file.

> Feature 2: input: --m-evenness [P], where P is optional and is a percentile of myelin thickness distribution around axon (default 10) output: in morphometrics files add _three_ columns: mean and std of myelin thickness distribution, plus Pth percentile of this distribution (if P is set to 0, i.e. a min value is requested, zero value will mean that axon is not complete, sanity check for features 1 and 2)
> 
> I guess both options can be combined into one flag with two optional parameters, -m [N] [P]
> 
> It is obviously up to you how you set it up, thank you for consideration.

@akraskovSMDB thank you for your input. In my head, it would make more sense to have 2 separate optional parameters, but these features are intertwined because the granularity of the angle increment will be used for both computations.

A cleaner way would be to have a flag that controls the granularity (let's call it `-A` for now). If this optional argument is provided, both myelin completeness and evenness would be computed. The percentile will default to 0 (min value). An additionnal optional `-B` flag would control the percentile value. If `-A` is present but not `-B`, the myelin evenness will still be computed. I think this would be more user-friendly (simpler for newer users) and give as much control to the user. We could go further and always output the minimum value, but the `-B` flag would add an additionnal column.

**Comment by hermancollin** [comment]
@akraskovSMDB during our meeting last week, we discussed adding these metrics and we are interested in working on this problem. Additionally, could you share the pixel size of the image you provided here? This is for internal archiving and I'll also probably be using this image to test the features during development.

The feedback you provide is very useful for us. Most users don't really interact with the dev team on github. If you find the time, we would be interested in hearing briefly the nature of your study and your use of ADS. You can email us directly at axondeepseg@googlegroups.com (we had problems in the past with external users sending us emails but it should work now).

**Reply by akraskovSMDB** [reply]
pixel size is 0.064um/pix, these are LM images from Firmin et al 2014, I will send an original one to axondeepseg address since the one attached to the post probably has different number of pixels

**Reply by akraskovSMDB** [reply]
> @akraskovSMDB during our meeting last week, we discussed adding these metrics and we are interested in working on this problem. 

Just wanted to check whether you have any time estimates on when the completeness and evenness of the myelin features can be implemented?
